### PR TITLE
Date and datetime

### PR DESCRIPTION
**Pull request by randomstuff**
Rename `Date` into `DateTime` and add `CalendarDate` (for `date`).

Same for `DateRange`.

For compatiblity `Date` is an alias for `DateTime`. The idea is to rename `CalendarDate` into `Date` later on.

see #346 and #168

### CODE REVIEW

**Comment by philippjfr** [comment]
This looks great to me. @jbednar @jlstevens Any objections?

**Comment by randomstuff** [comment]
(Merged the first two commits)

File: param/__init__.py [hunk 1]
```diff
@@ -1940,13 +1970,42 @@ def _validate(self, val):
...
+
+# Compatibility:
+DateRange = DateTimeRange
+
+
+class CalendarDateRange(Range):
+    """
+    A date range specified as (start_date, end_date).
+
+    Dates must be specified as datetime-like types (see
```
Comment (jbednar) [inline param/__init__.py]: Should this be "datetime" or "date"? The latter seems correct to me, but if the former is true, it's fine.

**Comment by randomstuff** [comment]
If this change is merges I'm not sure it makes sense to allow `DateTimeRange` to hold a `date` value instead of a `datetime` one. I find it's preddy strange:

```python
import datetime

import param

class Foo(param.Parameterized):
    p = param.DateTime()

foo = Foo()
foo.p = datetime.date.today()
print(repr(foo.p))
# => datetime.date(2019, 7, 25)
```

Changing this would be a breaking change however.

**Comment by randomstuff** [comment]
Fixed the documentation of the changes which has some confusing regarding datetime vs date, etc.

**Comment by randomstuff** [comment]
Another solution would be to keep `Date` dand `DateRange` for now (with the idea or replacing them later on with `CalendarDate` and `CalendarDateTime`) which would support both for now and add new `DateTime` and `DateTimeRange` which would only accept `datetime`

**Comment by randomstuff** [comment]
Updated to this alternative proposition:

* keep `Date` unmodified;
* `CalendarDate` for `date`-only.

I'd be tempted to add (in other PRs):

* [`NaiveDateTime`](https://www.w3.org/TR/html52/sec-forms.html#local-date-and-time-state-typedatetimelocal) for naive date times (no timezone information);
* `DateTime` (?) for non-naive datetime;
* [`Time`](https://www.w3.org/TR/html52/sec-forms.html#time-state-typetime) for (naive) `time`.

**Comment by philippjfr** [comment]
Thanks @randomstuff! Sorry for the long review cycle. This is hugely appreciated.

**Comment by randomstuff** [comment]
I was looking of the different `panel` widgets and how they would behave if using `CalendarDate` instead of `Date` this change and I discovered that `datetime` is actually a super class of `date`. (Which is not actually super useful but whatever …).

As a consequence, `CalendarDate` can currenly hold `datetime` values which is not what I intended. The only difference between `Date` and `CalendarDate` is that the latter cannot hold a `np.datetime64` value (because `np.datetime64` is not a super-class of `date`). This is somewhat nice because `np.datetime64` does not behave as a `date` at all (no `.month`, etc.).

Conclusion:

* As `datetime` is a superclass of `date`, it is kind-of expected that the `Date`/`CalendarDate` parameters would accept `datetime` values so the `Date` parameter type is actually OK in this regard. I wonder if it would not be useful to automatically convert `datetime` to `date` on assignment (either for `Date`/`CalendarDate`).

*  I'm not sure `CalendarDate` is actually any useful in its current state and I'd vote for reverting it for now (at least before release if no clear plan as been found).

* Allowing `Date` parameters to hold `np.datetime64` is not super useful because `np.datetime64` do not behave as `date` at all. It might be more useful to convert `np.datetime64` to `datetime`/`date` automatically when assigning.

**Pull request merged**